lib/ProductOpener/Tags.pm

```diff
@@ -344,6 +364,187 @@ sub get_inherited_property ($tagtype, $canon_tagid, $property) {
+						$found_properties{$property} = $property_value;
+						delete $unfound_properties{$property};
+					}
+				}
+			}
+			# add parents to the search ?
+			my $propagate = 0;
+			if (exists $direct_parents{$tagtype}{$tagid}) {
+				if (!defined $unfound_properties{$tagid}) {
+					$propagate = scalar %unfound_properties;
+				}
+				else {
+					# check if we have at least one unfonud property which not "undef"
+					for my $property (keys %unfound_properties) {
+						if (!defined $unfound_properties{$tagid}{$property}) {
+							$propagate = 1;
+							last;
+						}
+					}
+				}
+			}
+
+			if ($propagate) {
+				# propagate search to parents
+				foreach my $parent (keys %{$direct_parents{$tagtype}{$tagid}}) {
+					if (!defined $seen{$parent}) {
+						$seen{$parent} = 1;
+						push @parents, [$depth + 1, $parent];
+					}
+					# propagate undef, we merge with maybe existing items
+					if (defined $undef_properties{$tagid}) {
+						defined $undef_properties{$parent} or $undef_properties{$parent} = {};
+						foreach my $property (keys %{$undef_properties{$tagid}}) {
+							$undef_properties{$parent}{$property} = 1;
+						}
+					}
+				}
+
+				# sort parents, lower depth first, name second
+				@parents = sort {(@$a[0] <=> @$b[0]) || (@$a[1] cmp @$b[1])} @parents;
+			}
+
+			# no need to keep undef_properties for $tagid
+			delete $undef_properties{$tagid} if defined $undef_properties{$tagid};
+		}
+	}
+	return \%found_properties;
+}
+
+=head2 tags_by_prop ($tagtype, $tagids_ref, $prop_name, $props_ref, $inherited_props_ref, $fallback_lcs = ["xx", "en"])
+Retrieve properties of a series of tags given in C<$tagids_ref>
+and return them, but grouped by C<$prop_name>,
+also fetching C<$props_ref> and C<$inherited_props_ref>
+
+=head3 Return
+A ref to a hashmap, where keys are property C<$prop_name> values,
+and values are in turn hashmaps where keys are tag ids,
+and values are a hashmap with of properties and their values.
+
+Tags with undefined property are with group under "undef" value.
+
+=head4 Example
+we asks for quality tags, grouped by fix_action, while getting descriptions
+{
+	"add_nutrition_facts" => {
+		"en:kcal-does-not-match-other-nutrients" => {
+			"description:en" => "Kcal is not matching value computed from other nutriments"
+		},
+		"en:kcal-does-not-match-kj" => {
+			"description:en" => "Kcal is not matching kj value"
+		},
+	},
+	"add_categories" =>
+	{
+		"en:detected-category-baby-milk" {
+			"description:en" => "Detected category … may be missing baby milks"
+		}
+	}
+}
+=cut
+
+sub tags_by_prop ($tagtype, $tagids_ref, $prop_name, $props_ref, $inherited_props_ref, $fallback_lcs = ["xx", "en"]) {
+	my @tagids = @{$tagids_ref};
+	my @props_to_fetch = (@{$inherited_props_ref});
+	push @props_to_fetch, $prop_name;
+
+	my $tags_by_prop = {};
+
+	foreach my $tagid (@tagids) {
+		my $found_ref = get_inherited_properties($tagtype, $tagid, \@props_to_fetch);
```

Comment:
I think the code for tags_by_prop would be simpler to understand if we just called $prop_value = get_inherited_property($tagtype, $tagid, $prop_name), and then had a for each loop for each property we want, just like for the properties with fallback lcs.